### ISSUE TITLE
Binaries produced by urweb are not reproducible

### ISSUE THREAD

**Issue by bbarenblat** [urweb/urweb#38]
When `urweb` compiles a project, it emits code which contains datestamps in several places. In a trivial project,
1. the JavaScript runtime is served from a file named with a datestamp and
2. requests’ `If-modified-since` headers get checked against compilation time to determine whether or not to return a 304 Not Modified.

This means that Ur/Web applications cannot be built reproducibly – `urweb` will emit different binaries for the same code if compiled at different times.

Fixing these issues is relatively easy.
1. The compiler should hash the JavaScript runtime data and serve from a hash-keyed URL, not a date-keyed one.
2. Instead of using the compile time as the modification time of the runtime, the compiler should use the latest modification time of any file in the project.

**Comment by achlipala** [comment]
All sounds quite reasonable.  Are you offering to implement it? ;-)

**Comment by bbarenblat** [comment]
I can do it, though I don’t have a timeline in mind right now. This could also be a very good bug for somebody who’s new to the project and wants to get their feet wet with SML.

Why don’t we leave it unassigned for the time being? I’ll assign it to myself if I start working on it.

**Comment by jmitchell** [comment]
I'd like to get familiar with the code and give this a shot.

The contents of the `app.<compile-time>.js` file are determined by `JsComp.process` and later emitted to the `static char jslib[]` C array in `CjrPrint`. The `static void uw_handle` function emitted to the same C file handles all the HTTP 304 logic.

**Open questions:**

1. Is there a preferred string hashing algorithm to use on the `jslib` contents to name the JS file? Does it matter whether the hash function is implemented in C or SML?
2. Is the transitive closure of source files reachable from the initial `.urp` readily available? If not, is a conservative heuristic like the greatest `OS_FILE_SYS.modTime` in the entire directory tree suitable for this iteration?

**Comment by achlipala** [comment]
Thanks for offering to take on this task!

1. I imagine any reasonable algorithm is fine.  To me it is most natural to run the hashing algorithm at compile time and encode the result in the generated binary, which means that SML is the natural choice.
2. To compute the latest modification time: my suggestion is to instrument all file-open operations in the compiler, making them also look up the modification time and modify a global variable with a `max` operation.  File-open operations could come from this Cartesian product: {`BinIO`, `TextIO`}.{`openIn`, `openOut`}.  (Actually, `openOut` shouldn't be relevant here, but maybe something brings it into play.)

**Comment by bbarenblat** [comment]
For what it’s worth, it looks like [tom7](https://github.com/tom7) has written public-domain implementations of [MD5](https://github.com/robsimmons/sml-lib/tree/master/crypt/MD5) and [SHA-1](https://github.com/robsimmons/sml-lib/tree/master/crypt/SHA1).

**Comment by jmitchell** [comment]
Thanks for the tips! I've made progress on part 2 (https://github.com/jmitchell/urweb/commit/51ac19f565fa935eec6d9dd1f7119c6227383a3c). Based on some manual testing the caching works as expected. One handy test is adding `file /index.html index.html` to `hello.urp` and going through edit/compile/run cycles.

To avoid mtimes of files generated by the compiler affecting the `Last-Modified` value, I track when Ur/Web is reset. There may be room for improvement if we exclude generated files, e.g. in `/tmp`. On filesystems with 1-second mtime resolution, files generated soon after running `urweb` can set the `Last-Modified` time more conservatively than necessary.

I'll wait to file a pull request until I've finished part 1. Happy to take feedback on what's there now too.

**Comment by jmitchell** [comment]
I was about to integrate @tom7's MD5 implementation, but realized there's a potential licensing issue. [sml-lib's license](https://github.com/robsimmons/sml-lib/blob/master/COPYING) is GPLv3 and Ur/Web is released under BSD 3-clause. Comments in the MD5 source files suggest they may have originally been released as "public domain", so maybe if we can track down the original and/or clarify its license there's no problem.

Before I learned of @tom7's work I was considering implementing CRC. I could still do that if there are licensing concerns, although it will take a little longer.

**Comment by jmitchell** [comment]
Here's the original on [SourceForge](https://sourceforge.net/p/tom7misc/svn/HEAD/tree/trunk/sml-lib/). Same licensing story there.

**Comment by bbarenblat** [comment]
IANAL, but I feel comfortable copying a public domain file from a copyleft project into a permissive project, provided the file doesn’t depend on any copyleft code.

If you’d like to do your own implementation, I strongly recommend against a CRC; CRCs are designed for error detection, not file identification. Consider using a cryptographic hash function, like SHA-256.

**Comment by achlipala** [comment]
Have you tried asking @tom7 to be sure?  I might be able to help get his attention, if needed.

**Comment by tom7** [comment]
I wrote md5-sig.sml and md5.sml, and intend for those files to be placed in
the public domain. They only depend on the SML basis. There's also an
implementation of SHA-1 in sml-lib, which is a stronger algorithm (it has
been practical to generate full MD5 collisions for about a decade now).
Same deal with SHA-1: Public domain, standalone. Good luck!

 - Tom 7

On Thu, Sep 8, 2016 at 6:25 AM, Benjamin Barenblat <notifications@github.com

**Comment by jmitchell** [comment]
Thanks for clarifying, @tom7. I used your SHA-1 implementation in the submitted pull request.

**Comment by jmitchell** [comment]
I believe this can be closed now. The pull request has been merged into master.

**Comment by bbarenblat** [comment]
I’m still getting Last-Modified differences.

```
$ urweb
The Ur/Web compiler, version 20160805 + 3afe16344c02881e4ace9e8965f48154b5225638
$ echo 'val main = return <xml>Hi</xml>' >test.ur
$ urweb -output left test
$ urweb -output right test
$ diff -u <(strings left) <(strings right)
--- /dev/fd/63  2016-09-19 18:38:35.144311076 -0400
+++ /dev/fd/62  2016-09-19 18:38:35.144311076 -0400
@@ -1,4 +1,5 @@
 /lib64/ld-linux-x86-64.so.2
+1wz 
 liburweb_http.so.0
 uw_Basis_requestHeader
 __gmon_start__
@@ -44,7 +45,7 @@
 HTTP/1.1 404 Not Found
 Status: 404 Not Found
 If-modified-since
-Mon, 19 Sep 2016 22:36:42 GMT
+Mon, 19 Sep 2016 22:36:45 GMT
 /Stuff/main
 <body
 >Hi</body>
@@ -53,7 +54,7 @@
 Not Found
 /app.DA39A3EE5E6B4B0D3255BFEF95601890AFD80709.js
 Content-Type: text/javascript
-Last-Modified: Mon, 19 Sep 2016 22:36:42 GMT
+Last-Modified: Mon, 19 Sep 2016 22:36:45 GMT
 Cache-Control: max-age=31536000, public
 Content-type: text/html; charset=utf-8
 ;*3$"
```

Neither of the Last-Modified dates is the mtime of test.ur.

**Comment by jmitchell** [comment]
I can take a closer look later, but I suspect the temporary generated files and the filesystem's limited modtime resolution are causing this. If so, we could improve it by not just tracking the compiler start time, but also the list of files the compiler modifies.

**Comment by jmitchell** [comment]
@bbarenblat, thanks for catching this. The problem was as I expected. I've submitted a new pull request with a fix.

Rather than maintaining a list of files modified by the compiler I opted to reduce the time comparisons to 1s precision. Generally filesystems have 1s or better mtime resolution and `Last-Modified` also uses 1s precision, so this should be fairly reliable. I prefer this approach over the modified file list because such a list could be difficult to maintain as the codebase evolves.

**Comment by bbarenblat** [comment]
This still won’t work on FAT, which has two-second mtime precision. Why is it difficult to differentiate between files the compiler takes as input and files the compiler generates?

**Comment by vshabanov** [comment]
Maybe it worth to set fixed `Last-Modified` date or even use `ETag` instead?

If I build the same app on different servers file modification time will differ anyway.

**Comment by achlipala** [comment]
I think I can help understand the places where the compiler reads its own file-system writes, so that they can be disregarded.

First, I will assume that we only care about reproducibility for successful builds.  (Generation of error messages sometimes involves reading from self-written files.  Don't ask. ;-])

Then that leaves a few other spots:
1. `compiler.sml`:`parseUrs`: the code `FileIO.txtOpenIn` should change back to `TextIO.openIn`, since here we read a file written a few lines above.
2. `tutorial.sml`: some offenders here, but do we mind if tutorial building isn't reproducible?
3. `demo.sml`: `TextIO.openOut` for `"demo.urs"`, `"demo.ur"`, and "`demo.urp`".  These are extra source files created for compilation of demos that combine multiple apps.  The current `make test` happens to use the demo mode, so it may be worth catching these cases.  How about extending `FileIO` with a mutable whitelist of files whose timestamps should be ignored, adding to that list only in the three spots I've mentioned in `demo.sml`?  (Might as well do the same for tutorials, then, I suppose.)

**Comment by jmitchell** [comment]
I can revisit this in a few days.

**Comment by bbarenblat** [comment]
This seems to be fixed in 20161022.
